src/tar_parser.rs

```diff
@@ -0,0 +1,246 @@
+//! A This module implements a parser for uncompressed tar files,
+//! including the ustar extension (but not yet supporting the PAX extension).
+
+use core::ops::Range;
+
+use crate::filename::{truncate, Filename};
+
+/// A parser for an already decompressed image / tar ball
+#[derive(Clone, Copy)]
+pub struct Parser<'a> {
+    input: &'a BytesSlice,
+    offset: usize,
+}
+
+impl<'a> Parser<'a> {
+    /// Constructs a new parser from a given tar byte slice
+    pub fn new(input: &'a BytesSlice) -> Self {
+        Self { input, offset: 0 }
+    }
+}
+
+/// Metadata for a tar file entry.
+#[derive(Clone, Copy, Debug, Default, PartialEq, Eq)]
+pub struct FileMetadata {
+    /// Is the file entry marked as executable?
+    pub is_exec: bool,
+}
+
+/// A file entry in a tar ball
+#[derive(Clone, Debug)]
+pub struct File<'a> {
+    /// The full file name of the entry
+    pub name: Filename<'a>,
+
+    /// The associated file metadata
+    pub metadata: FileMetadata,
+
+    /// The indices range of the value associated to this entry
+    pub value_range: Range<usize>,
+
+    /// A slice of the value (file content) associated to this entry
+    pub value: &'a [u8],
+}
+
+/// Possible parser errors:
+#[derive(Clone, Debug, PartialEq)]
+pub enum ParserError<'a> {
+    /// The parser encountered the end of the file during parsing of an entry.
+    UnexpectedEof,
+
+    /// The parser encountered something which was supposed to be an integer, but wasn't.
+    ParseInt(core::num::ParseIntError),
+
+    /// The parser encountered an integer which was out-of-range.
+    FromInt(core::num::TryFromIntError),
+
+    /// The parser encountered something which
+    /// was supposed to be an UTF-8 string (usually an ASCII integer), but wasn't.
+    Utf8(core::str::Utf8Error),
+
+    /// The thin tree encountered a file which got overridden by a directory.
+    FileOverridenWithDirectory(Filename<'a>),
+}
+
+impl<'a> From<core::num::ParseIntError> for ParserError<'a> {
+    #[inline(always)]
+    fn from(x: core::num::ParseIntError) -> Self {
+        Self::ParseInt(x)
+    }
+}
+
+impl<'a> From<core::num::TryFromIntError> for ParserError<'a> {
+    #[inline(always)]
+    fn from(x: core::num::TryFromIntError) -> Self {
+        Self::FromInt(x)
+    }
+}
+
+impl<'a> From<core::str::Utf8Error> for ParserError<'a> {
+    #[inline(always)]
+    fn from(x: core::str::Utf8Error) -> Self {
+        Self::Utf8(x)
+    }
+}
+
+#[cfg(feature = "alloc")]
+/// Owned bytes belonging to a decompressed tar file
+#[derive(Clone, Default, PartialEq, Eq)]
+pub struct Bytes(pub alloc::boxed::Box<[u8]>);
```

Comment:
```suggestion
pub struct TarBytes(pub alloc::boxed::Box<[u8]>);
```

Maybe this makes the purpose of the struct a little easier to gasp.

Reply:
True, but `Bytes` is a _very_ generic name. It kind of forces the use of `tar_parser::Bytes` to achieve easily comprehendible code.

Reply:
Note that we rarely if ever have to write that type: we only write it out once in https://github.com/hermit-os/uhyve/pull/1135.